File: backend/internal/domain/residente/membresia_model.go
// backend/internal/domain/residente/membresia_model.go
package residente

import "gorm.io/gorm"

const (
	// Reusadas por Membresia.Status (y antes por Residente.Status, ya
	// eliminado) — el ciclo de aprobación activo/pendiente/rechazado es el
	// mismo para ambos.
	ResidenteStatusActivo    = "activo"
	ResidenteStatusPendiente = "pendiente"
	ResidenteStatusRechazado = "rechazado"

	// Membresia.Rol -- RolInvitadoFrecuente reusa exactamente el mismo
	// mecanismo de reconocimiento facial que un residente (cualquier
	// Membresia activa con embedding entra por match de rostro, ver
	// persona.Repository.FindActivasPorTenant) en vez de un sistema de
	// reconocimiento paralelo. Lo que cambia es semántico: no es
	// "residente" de la casa, es un invitado al que un residente le dio
	// acceso recurrente y puede revocar cuando quiera.
	RolResidente         = "residente"
	RolInvitadoFrecuente = "invitado_frecuente"
)

// Membresia es la relación de una Persona (internal/domain/persona) con un
// tenant específico — reemplaza lo que hoy vive dentro de Residente (nombre,
// teléfono y embedding se mueven a Persona; casa_destino, pin y rol se
// quedan aquí, porque son propios de cada centro habitacional). Ver spec
// 2026-08-16-persona-identidad-kigo-design.md §2.
//
// Status reusa las constantes ResidenteStatus* de arriba (activo |
// pendiente | rechazado) — el ciclo de aprobación es el mismo.
type Membresia struct {
	gorm.Model
	PersonaID   uint   `gorm:"column:persona_id;not null;index"`
	TenantID    uint   `gorm:"column:tenant_id;not null;index"`
	CasaDestino string `gorm:"not null"`
	// DestinoID liga con destinos.id cuando CasaDestino pudo resolverse a un
	// Destino real -- null en membresías viejas sin match.
	DestinoID *uint `gorm:"column:destino_id;index"`
	// Pin guarda el hash bcrypt (lo que compara el kiosko, incluso
	// offline); PinCodigo guarda los 5 dígitos en claro que genera el
	// backend al crear la membresía, porque la app se los tiene que poder
	// mostrar al residente en "Mi QR". El PIN ya no lo elige la persona y
	// no cambia una vez asignado.
	Pin                         string `gorm:"not null"`
	PinCodigo                   string `gorm:"column:pin_codigo;not null;default:''"`
	Status                      string `gorm:"not null;default:'activo'"`
	PermiteReconocimientoFacial bool   `gorm:"not null;default:false"`
	KioskoID                    *uint
	TiempoEsperaMin             *int
	// Rol distingue un residente real de un invitado frecuente -- ver
	// RolInvitadoFrecuente arriba. Default 'residente' para no romper filas
	// existentes.
	Rol string `gorm:"not null;default:'residente'"`
}

func (Membresia) TableName() string { return "membresias" }

File: kigo-app/lib/models/membresia_model.dart
/// Una fila de la respuesta de GET /personas/me/membresias — la Membresia
/// más reciente de la Persona (se asume una sola activa a la vez, ver
/// spec 2026-08-17-kigo-app-rediseno-design.md §8).
class MembresiaActual {
  final int id;
  final int tenantId;
  final String centroNombre;
  final String casaDestino;
  final String status;
  // 'residente' o 'invitado_frecuente' -- un invitado frecuente entra por
  // rostro/QR, no vive en el centro. La UI no debe presentárselo como si
  // fuera un residente inscrito (ver settings_view.dart).
  final String rol;

  bool get esResidente => rol == 'residente';

  /// PIN de 5 dígitos que genera el backend al crear la membresía — la
  /// persona ya no lo elige y no cambia. Se muestra en "Mi QR". Vacío para
  /// un invitado frecuente (entra por rostro/QR, no por PIN).
  final String pin;

  MembresiaActual({
    required this.id,
    required this.tenantId,
    required this.centroNombre,
    required this.casaDestino,
    required this.status,
    this.rol = 'residente',
    required this.pin,
  });

  factory MembresiaActual.fromJson(Map<String, dynamic> json) {
    return MembresiaActual(
      id: json['id'] as int,
      tenantId: json['tenant_id'] as int,
      centroNombre: json['centro_nombre'] as String? ?? '',
      casaDestino: json['casa_destino'] as String,
      status: json['status'] as String,
      rol: json['rol'] as String? ?? 'residente',
      pin: json['pin'] as String? ?? '',
    );
  }
}
